File: 3235-check-if-the-rectangle-corner-is-reachable/3235-check-if-the-rectangle-corner-is-reachable.java
class Solution {

    public boolean canReachCorner(int X, int Y, int[][] circles) {
        ArrayDeque<int[]> deque = new ArrayDeque<>();
        for (int i = 0; i < circles.length; i++) {
            if (circles[i][0] - circles[i][2] <= 0 || circles[i][1]  + circles[i][2] >= Y) {
                deque.offer(circles[i]);
                circles[i] = null;
            }
        }
        for (int[] circle; !deque.isEmpty();) {
            if ((circle = deque.poll())[0] + circle[2] >= X || circle[1] - circle[2] <= 0) {
                return false;
            }
            for (int i = 0; i < circles.length; i++) {
                if (circles[i] != null && Math.abs(circle[0] - circles[i][0]) * Math.abs(circle[0] - circles[i][0]) + Math.abs(circle[1] - circles[i][1]) * Math.abs(circle[1] - circles[i][1]) <= (circle[2] + circles[i][2]) * (circle[2] + circles[i][2])) {
                    deque.offer(circles[i]);
                    circles[i] = null;
                }
            }
        }
        return true;
    }
}

File: 3235-check-if-the-rectangle-corner-is-reachable/3235-check-if-the-rectangle-corner-is-reachable.py
class Solution(object):
    def canReachCorner(self, X, Y, circles):
        """
        :type X: int
        :type Y: int
        :type circles: List[List[int]]
        :rtype: bool
        """
        n = len(circles)
        L, R, D, U = [], [], [], []
        markL = [False] * n
        markR = [False] * n
        markD = [False] * n
        markU = [False] * n
        mark = [False] * n
        for i in range(n):
            x, y, r = circles[i]
            cnt = 0
            if x - r <= 0:
                cnt += 1
                markL[i] = True
                L.append(i)
            if x + r >= X:
                cnt += 1
                markR[i] = True
                R.append(i)
            if y - r <= 0:
                cnt += 1
                markD[i] = True
                D.append(i)
            if y + r >= Y:
                cnt += 1
                markU[i] = True
                U.append(i)
        q = deque()
        for v in L:
            mark = [False] * n
            q.append(v)
            mark[v] = True
            while q:
                v = q.popleft()
                if markR[v] or markD[v]:
                    return False
                for i in range(n):
                    x1, y1, r1 = circles[v]
                    x2, y2, r2 = circles[i]
                    if not mark[i] and self.conflict(x1, y1, r1, x2, y2, r2):
                        q.append(i)
                        mark[i] = True
        for v in U:
            mark = [False] * n
            q.append(v)
            mark[v] = True
            while q:
                v = q.popleft()
                if markR[v] or markD[v]:
                    return False
                for i in range(n):
                    x1, y1, r1 = circles[v]
                    x2, y2, r2 = circles[i]
                    if not mark[i] and self.conflict(x1, y1, r1, x2, y2, r2):
                        q.append(i)
                        mark[i] = True
        return True
    def conflict(self, x1, y1, r1, x2, y2, r2):
        return (abs(x1 - x2) ** 2 + abs(y1 - y2) ** 2 <= (r1 + r2) ** 2)